runtime/src/main/java/org/corfudb/runtime/object/CorfuCompileProxy.java

```diff
@@ -302,7 +302,13 @@ public UUID getStreamID() {
         int retries = 1;
         while (true) {
             try {
-                rt.getObjectsView().TXBegin();
+                /* If it is a nested transaction, inherit type of parent */
+                if (TransactionalContext.getCurrentContext() != null) {
```

Comment:
I think this should be fixed in a higher layer (i.e. ObjectsView class).

Consider something like:


@@ -105,8 +106,14 @@ public class ObjectsView extends AbstractView {
      * to other threads or clients until TXEnd is called.
      */
     public void TXBegin() {
+        TransactionType type = TransactionType.OPTIMISTIC;
+
+        if(TransactionalContext.isInTransaction()) {
+            type = TransactionalContext.getCurrentContext().getBuilder().type;
+        }
+
         TXBuild()
-                .setType(TransactionType.OPTIMISTIC)
+                .setType(type)
                 .begin();
     }